
---
title: "Inference and analysis of cell-cell communication using CellChat"
author: "Suoqin Jin"
output: html_document
mainfont: Arial
vignette: >
  %\VignetteIndexEntry{Inference and analysis of cell-cell communication using CellChat}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = './'
)
knitr::opts_chunk$set(eval = FALSE)
```


This vignette outlines the steps of inference, analysis and visualization of cell-cell communication network using CellChat. 

CellChat requires gene expression data of cells as the user input and models the probability of cell-cell communication by integrating gene expression with prior knowledge of the interactions between signaling ligands, receptors and their cofactors. 

Upon infering the intercellular communication network, CellChat provides functionality for further data exploration, analysis, and visualization. 


## Load the required libraries
```{r message=FALSE,warning=FALSE}
library(CellChat)
library(ggplot2)
library(ggalluvial)
options(stringsAsFactors = FALSE)
```

# Part I: Data input & processing and initialization of CellChat object
CellChat requires two user inputs: one is the gene expression data of cells, and the other is either user assigned cell labels (i.e., label-based mode) or a low-dimensional representation of the single-cell data (i.e., label-free mode). For the latter, CellChat automatically groups cells by building a shared neighbor graph based on the cell-cell distance in the low-dimensional space or the pseudotemporal trajectory space. 

## Load data
*For the gene expression data matrix*, genes should be in rows with rownames and cells in columns with colnames. Normalized data is required as input for CellChat analysis. If user provides count data, we provide a `normalizeData` function to account for library size. *For the cell group information*, a dataframe with rownames is required as input for CellChat. 

```{r}
load("data_input.rda")
data.input = data_input$data # normalized data matrix
identity = data.frame(group = data_input$labels, row.names = names(data_input$labels)) # create a dataframe consisting of the cell labels
unique(identity$group) # check the cell labels
```

## Create a CellChat object
```{r}
cellchat <- createCellChat(data = data.input)
```

## Add cell information into *meta* slot of the object
```{r}
cellchat <- addMeta(cellchat, meta = identity, meta.name = "labels")
cellchat <- setIdent(cellchat, ident.use = "labels") # set "labels" as default cell identity
levels(cellchat@idents) # show factor levels of the cell labels
groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group
```

## Set the ligand-receptor interaction database
```{r}
CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on human data
showDatabaseCategory(CellChatDB)
# Show the structure of the database
dplyr::glimpse(CellChatDB$interaction)

CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") # use Secreted Signaling for cell-cell communication analysis
cellchat@DB <- CellChatDB.use # set the used database in the object
```

## Preprocessing the expression data for cell-cell communication analysis
We first identify over-expressed ligands or receptors in one cell group, and then project gene expression data onto protein-protein interaction (PPI) network. The over-expressed ligand-receptor interactions are identified if either the ligand or receptor is over-expressed. 
```{r}
cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
future::plan("multiprocess", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- projectData(cellchat, PPI.mouse)
```

# Part II: Inference of cell-cell communication network
CellChat infers the biologically significant cell-cell communication by assigning each interaction with a probability value and peforming a permutation test. CellChat models the probability of cell-cell communication by integrating gene expression with prior known knowledge of the interactions between signaling ligands, receptors and their cofactors using the law of mass action. 

## Compute the communication probability and infer cellular communication network
```{r}
cellchat <- computeCommunProb(cellchat)
```

## Infer the cell-cell communication at a signaling pathway level
CellChat computes the communication probability on signaling pathway level by summarizing the communication probabilities of all ligands-receptors interactions associated with each signaling pathway.  

NB: The inferred intercellular communication network of each ligand-receptor pair and each signaling pathway is stored in the slot 'net' and 'netP', respectively.

```{r}
cellchat <- computeCommunProbPathway(cellchat)
```

## Calculate the aggregated cell-cell communication network 
We can calculate the aggregated cell-cell communication network by counting the number of links or summarizing the communication probability
```{r}
cellchat <- aggregateNet(cellchat)
```

# Part III: Visualization and systems analysis of cell-cell communication network
Upon infering the cell-cell communication network, CellChat provides various functionality for further data exploration, analysis, and visualization.

* It can quantitatively characterize and compare the inferred cell-cell communication networks using an integrated approach by combining social network analysis, pattern recognition, and manifold learning approaches.

* It provides an easy-to-use tool for extracting and visualizing high-order information of the inferred networks. For example, it allows ready prediction of major signaling inputs and outputs for all cell populations and how these populations and signals coordinate together for functions.

## Create a directory to save figures
```{r}
data.dir <- '/Users/example/results'
dir.create(data.dir)
setwd(data.dir)
```

## Visualize each signaling pathway using hierarchy plot or circle plot
If visualizing the communication network using hierarchy plot, please define `vertex.receiver`, which is a numeric vector giving the index of the cell groups as targets in the first hierarchy plot. One can visualize the inferred communication network of signaling pathways using `netVisual_aggregate`, and visualize the inferred communication networks of individual L-R pairs associated with that signaling pathway using `netVisual_individual`.

In the hierarchy plot, solid and open circles represent source and target, respectively. Circle sizes are proportional to the number of cells in each cell group. Edge colors are consistent with the signaling source. Thicker line indicates a stronger signal. 

Here we take input of one signaling pathway as an example: 
```{r}
pathways.show <- c("TGFb") 
vertex.receiver = seq(1,9) # a numeric vector
# Hierarchy plot
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver, vertex.size = groupSize)
# Circle plot
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle", vertex.size = groupSize)
```

## Compute and visualize the contribution of each ligand-receptor pair to the overall signaling pathway
```{r}
netAnalysis_contribution(cellchat, signaling = pathways.show)
```

## Identify signaling roles of cell groups
CellChat allows ready identification of dominant senders, receivers, mediators and influencers in the intercellular communication network by computing several network centrality measures for each cell group.

```{r}
cellchat <- netAnalysis_signalingRole(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
```

## Visualize the signaling roles of cell groups
```{r}
netVisual_signalingRole(cellchat, signaling = pathways.show)
```

## Identify global communication patterns and major signals for specific cell groups
In addition to exploring detailed communications for individual pathways, an important question is how multiple cell groups and signaling pathways coordinate to function. CellChat employs a pattern recognition method to identify the global communication patterns as well as the key signals for each cell group. 

### Identify and visualize outgoing communication pattern of secreting cells
As the number of patterns increases, there might be redundant patterns, making it difficult to interpret the communication patterns. We chose five patterns as default. Generally, it is biologically meaningful with the number of patterns greater than 2. 
```{r}
nPatterns = 5
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
# river plot
netAnalysis_river(cellchat, pattern = "outgoing")
# dot plot
netAnalysis_dot(cellchat, pattern = "outgoing")
```

### Identify and visualize incoming communication pattern of target cells
```{r}
nPatterns = 5
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = nPatterns)
# river plot
netAnalysis_river(cellchat, pattern = "incoming")
# dot plot
netAnalysis_dot(cellchat, pattern = "incoming")
```

## Manifold and classification learning analysis of signaling networks
Further, CellChat is able to quantify the similarity between all significant signaling pathways and then group them based on their cellular communication network similarity. Grouping can be done either based on the functional or structural similarity. 

**Functional similarity**: High degree of functional similarity indicates major senders and receivers are similar, and it can be interpreted as the two signaling pathways or two ligand-receptor pairs exhibit similar and/or redundant roles. 

**Structural similarity**: A structural similarity was used to compare their signaling network structure, without considering the similarity of senders and receivers.

### Identify signaling groups based on their functional similarity
```{r}
cellchat <- computeNetSimilarity(cellchat, type = "functional")
cellchat <- netEmbedding(cellchat, type = "functional")
cellchat <- netClustering(cellchat, type = "functional")
# Visualization in 2D-space
netVisual_embedding(cellchat, type = "functional")
netVisual_embeddingZoomIn(cellchat, type = "functional")
```

### Identify signaling groups based on structure similarity
```{r}
cellchat <- computeNetSimilarity(cellchat, type = "structural")
cellchat <- netEmbedding(cellchat, type = "structural")
cellchat <- netClustering(cellchat, type = "structural")
# Visualization in 2D-space
netVisual_embedding(cellchat, type = "structural")
netVisual_embeddingZoomIn(cellchat, type = "structural")
```

## Part IV: Save the CellChat object
```{r}
saveRDS(cellchat, file = "cellchat_example.rds")
```

