
---
title: "Inference and analysis of cell-cell communication using CellChat"
author: "Suoqin Jin"
output: html_document
mainfont: Arial
vignette: >
  %\VignetteIndexEntry{Inference and analysis of cell-cell communication using CellChat}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = './'
)
knitr::opts_chunk$set(eval = FALSE)
```


This vignette outlines the steps of inference, analysis and visualization of cell-cell communication network using CellChat. 

CellChat requires gene expression data of cells as the user input and models the probability of cell-cell communication by integrating gene expression with prior knowledge of the interactions between signaling ligands, receptors and their cofactors. 

Upon infering the intercellular communication network, CellChat provides functionality for further data exploration, analysis, and visualization. 


Load the required libraries
```{r message=FALSE,warning=FALSE}
library(CellChat)
library(ggplot2)
library(ggalluvial)
options(stringsAsFactors = FALSE)
```

# Part I: Data input & processing and initialization of CellChat object
CellChat requires two user inputs: one is the gene expression data of cells, and the other is either user assigned cell labels (i.e., label-based mode) or a low-dimensional representation of the single-cell data (i.e., label-free mode). For the latter, CellChat automatically groups cells by building a shared neighbor graph based on the cell-cell distance in the low-dimensional space or the pseudotemporal trajectory space. 

## Load data
*For the gene expression data matrix*, genes should be in rows with rownames and cells in columns with colnames. Normalized data is required as input for CellChat analysis. If user provides count data, we provide a `normalizeData` function to account for library size. *For the cell group information*, a dataframe with rownames is required as input for CellChat. 

```{r}
load("data_input.rda")
data.input = data_input$data # normalized data matrix
identity = data.frame(group = data_input$labels, row.names = names(data_input$labels)) # create a dataframe of the cell labels
unique(identity$group) # show the cell labels
```

## Create a CellChat object
```{r}
cellchat <- createCellChat(data = data.input)
```

## Add cell information into *meta* slot of the object
```{r}
cellchat <- addMeta(cellchat, meta = identity, meta.name = "labels")
cellchat <- setIdent(cellchat, ident.use = "labels") # set "labels" as default cell identity
levels(cellchat@idents) # show factor levels of the cell labels
groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group
```

## Set the ligand-receptor interaction database
```{r}
CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on human data
showDatabaseCategory(CellChatDB)
# Show the structure of the database
dplyr::glimpse(CellChatDB$interaction)

CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") # use Secreted Signaling for cell-cell communication analysis
cellchat@DB <- CellChatDB.use # set the used database in the object
```

## Preprocessing the data for cell-cell communication analysis
We first identify over-expressed ligands/receptors and pairs, and then project gene expression data onto protein-protein interaction (PPI) network. 
```{r}
cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
future::plan("multiprocess", workers = 4) # parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- projectData(cellchat, PPI.mouse)
```

# Part II: Inference of cell-cell communication network
CellChat models the probability of cell-cell communication by integrating gene expression with prior known knowledge of the interactions between signaling ligands, receptors and their cofactors using the law of mass action. 
## Compute the communication probability and infer cellular communication network
```{r}
cellchat <- computeCommunProb(cellchat)
```

## Compute communications on signaling pathway level
CellChat computes the communication probability on signaling pathway level by summarizing all related ligands/receptors.
```{r}
cellchat <- computeCommunProbPathway(cellchat)
```

## Calculate the aggregated network 
We can calculate the aggregated network by counting the number of links or summarizing the communication probability
```{r}
cellchat <- aggregateNet(cellchat)
```

# Part III: Systems analysis and visualization of cell-cell communication network
Upon infering the cell-cell communication network, CellChat provides various functionality for further data exploration, analysis, and visualization.

## Create a directory to save figures
data.dir <- '/Users/example/results'
dir.create(data.dir)
setwd(data.dir)

## Identify signaling roles of cell groups
We identify dominant senders, receivers, mediators and influencers in the intercellular communication network by computing the several network centrality measures.

```{r}
cellchat <- netAnalysis_signalingRole(cellchat, slot.name = "netP")
```

## Visualize each signaling pathway using hierarchy plot or circle plot
Here we take input of one signaling pathway as an example: 
```{r}
pathways.show <- c("TGFb") 
vertex.receiver = seq(1,9)
# Hierarchy plot
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver, vertex.size = groupSize)
# Circle plot
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle", vertex.receiver = vertex.receiver, vertex.size = groupSize)
```

## Contribution of each L-R pair to overall signaling
```{r}
netAnalysis_contribution(cellchat, signaling = pathways.show)
```

## Visualize the signaling roles of cell groups using heatmap
```{r}
netVisual_signalingRole(cellchat, signaling = pathways.show)
```

## Identify and visualize outgoing communication pattern of secreating cells
```{r}
nPatterns = 5
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
# river plot
netAnalysis_river(cellchat, pattern = "outgoing")
# dot plot
netAnalysis_dot(cellchat, pattern = "outgoing")
```

## Identify and visualize incoming communication pattern of target cells
```{r}
nPatterns = 5
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = nPatterns)
# river plot
netAnalysis_river(cellchat, pattern = "incoming")
# dot plot
netAnalysis_dot(cellchat, pattern = "incoming")
```

## Manifold and classification learning analysis of signaling networks
### Compute and visualize functional similarity
```{r}
cellchat <- computeNetSimilarity(cellchat, type = "functional")
cellchat <- netEmbedding(cellchat, type = "functional")
cellchat <- netClustering(cellchat, type = "functional")
# Visualization in 2D-space
netVisual_embedding(cellchat, type = "functional")
netVisual_embeddingZoomIn(cellchat, type = "functional")
```

### Compute and visualize structural similarity
```{r}
cellchat <- computeNetSimilarity(cellchat, type = "structural")
cellchat <- netEmbedding(cellchat, type = "structural")
cellchat <- netClustering(cellchat, type = "structural")
# Visualization in 2D-space
netVisual_embedding(cellchat, type = "structural")
netVisual_embeddingZoomIn(cellchat, type = "structural")
```

# Part IV: Save the CellChat object
```{r}
saveRDS(cellchat, file = "cellchat_example.rds")
```

