
---
title: "Example streamline for quick analysis and exploration"
author: "Suoqin Jin"
output: html_document
mainfont: Arial
vignette: >
  %\VignetteIndexEntry{Inference and analysis of cell-cell communication using CellChat}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = './'
)
```
CellChat is an user-friendly tool that can fulfill the needs of users like biologists without much coding skills. In this example walkthrough, **we provide a streamline for quick analysis and exploration of cell-cell communication analysis, and will automatically save all the generated figures into an user-defined folder** (NB: User might need to set proper figure sizes for some plots). Such user-friendly way allows quick analysis and check of the inferred cell-cell communication networks and other advanced analysis.  

Here we apply CellChat to our own scRNA-seq data for day 12 mouse skin wound tissue from our recent paper [Single-cell analysis reveals fibroblast heterogeneity and myeloid-derived adipocyte progenitors in murine skin wounds](https://www.nature.com/articles/s41467-018-08247-x). We also demonstrate the package's visualization capabilities while reproducing the figure panels from the manuscript. 


## Load the required libraries
```{r message=FALSE,warning=FALSE}
library(CellChat)
library(ggplot2)
library(ggalluvial)
library(svglite)
options(stringsAsFactors = FALSE)
```

# I: Data input & processing

## Load data
CellChat requires two user inputs: one is the gene expression data of cells, and the other is either user assigned cell labels (i.e., label-based mode) or a low-dimensional representation of the single-cell data (i.e., label-free mode). Here we use the label-based mode by loading the assigned cell labels. 

*For the gene expression data matrix*, genes should be in rows with rownames and cells in columns with colnames. Normalized data is required as input for CellChat analysis. If user provides count data, we provide a `normalizeData` function to account for library size. *For the cell group information*, a dataframe with rownames is required as input for CellChat. 

The processed dataset, including data matrix and cell labels, can be downloaded from [here](https://www.dropbox.com/s/e8k6r3bkdr2twol/data_wound_CellChat.rda?dl=0). We first prepare the required input of CellChat. 


```{r}
load("/Users/suoqinjin/Downloads/data_wound_CellChat.rda")
data.input = data_wound$data # normalized data matrix
identity = data.frame(group = data_wound$labels, row.names = names(data_wound$labels)) # create a dataframe consisting of the cell labels
unique(identity$group) # check the cell labels
```

## Create a CellChat object
```{r}
cellchat <- createCellChat(data = data.input)
```

## Add cell information into *meta* slot of the object
```{r}
cellchat <- addMeta(cellchat, meta = identity, meta.name = "labels")
cellchat <- setIdent(cellchat, ident.use = "labels") # set "labels" as default cell identity
levels(cellchat@idents) # show factor levels of the cell labels
groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group
```

## Set the ligand-receptor interaction database
```{r}
CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on human data
CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") # use Secreted Signaling for cell-cell communication analysis
cellchat@DB <- CellChatDB.use # set the used database in the object
```

## Preprocessing the expression data for cell-cell communication analysis
We first identify over-expressed ligands or receptors in one cell group, and then project gene expression data onto protein-protein interaction (PPI) network. The over-expressed ligand-receptor interactions are identified if either the ligand or receptor is over-expressed. 
```{r,message=FALSE,warning=FALSE}
cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
future::plan("multiprocess", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- projectData(cellchat, PPI.mouse)
```

# II: Inference of cell-cell communication network
We then infer the biologically significant cell-cell communication by assigning each interaction with a probability value and peforming a permutation test. 

## Compute the communication probability and infer the cellular communication network
```{r}
cellchat <- computeCommunProb(cellchat)
```

## Infer the cell-cell communication at a signaling pathway level
We also compute the communication probability at a signaling pathway level by summarizing the communication probabilities of all ligands-receptors interactions associated with each signaling pathway.  

NB: The inferred intercellular communication network of each ligand-receptor pair and each signaling pathway is stored in the slot 'net' and 'netP', respectively.
```{r}
cellchat <- computeCommunProbPathway(cellchat)
```

## Calculate the aggregated cell-cell communication network 
We can calculate the aggregated cell-cell communication network by counting the number of links or summarizing the communication probabilities.
```{r}
cellchat <- aggregateNet(cellchat)
```

## Identify signaling roles of cell groups
CellChat allows ready identification of dominant senders, receivers, mediators and influencers in the intercellular communication network by computing several network centrality measures for each cell group.

```{r}
cellchat <- netAnalysis_signalingRole(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
```


# III: Visualization and systems analysis of cell-cell communication network
Upon infering the cell-cell communication network, CellChat provides various functionality for further exploration, analysis, and visualization.

* It can quantitatively characterize and compare the inferred cell-cell communication networks using an integrated approach by combining social network analysis, pattern recognition, and manifold learning approaches.

* It provides an easy-to-use tool for extracting and visualizing high-order information of the inferred networks. For example, it allows ready prediction of major signaling inputs and outputs for all cell populations and how these populations and signals coordinate together for functions.

## Create a directory to save figures
```{r,message=FALSE,warning=FALSE}
data.dir <- '/Users/suoqinjin/Documents/CellChat/walkthrough/example'
dir.create(data.dir)
setwd(data.dir)
```

## Visualize communication network of signaling pathway using hierarchy plot or circle plot
If visualizing the communication network using hierarchy plot, please define `vertex.receiver`, which is a numeric vector giving the index of the cell groups as targets in the first hierarchy plot. One can visualize the inferred communication network of signaling pathways using `netVisual_aggregate`, and visualize the inferred communication networks of individual L-R pairs associated with that signaling pathway using `netVisual_individual`.

In the hierarchy plot, solid and open circles represent source and target, respectively. Circle sizes are proportional to the number of cells in each cell group. Edge colors are consistent with the signaling source. Thicker line indicates a stronger signal. 

```{r}
# Access all the signaling pathways showing significant communications
pathways.show <- cellchat@netP$pathways
# check the order of cell identity to set suitable vertex.receiver
levels(cellchat@idents)
vertex.receiver = seq(1,9)
for (i in 1:length(pathways.show)) {
  # Visualize communication network associated with both signaling pathway and individual L-R pairs
  netVisual(cellchat, signaling = pathways.show[i], vertex.receiver = vertex.receiver, vertex.size = groupSize)
  # Compute and visualize the contribution of each ligand-receptor pair to the overall signaling pathway
  gg <- netAnalysis_contribution(cellchat, signaling = pathways.show[i])
  ggsave(filename=paste0(pathways.show[i], "_L-R_contribution.pdf"), plot=gg, width = 3, height = 2, units = 'in', dpi = 300)
  # Visualize signaling roles of cell groups
  grDevices::pdf(file = paste0("cellSignalingRole_",pathways.show[i],".pdf"))
  netVisual_signalingRole(cellchat, pathways.show[i], slot.name = "netP", width = 10)
  dev.off()
}
```
## Identify global communication patterns and major signals for specific cell groups
In addition to exploring detailed communications for individual pathways, an important question is how multiple cell groups and signaling pathways coordinate to function. CellChat employs a pattern recognition method to identify the global communication patterns as well as the key signals for each cell group. 

### Identify outgoing communication pattern of secreting cells
As the number of patterns increases, there might be redundant patterns, making it difficult to interpret the communication patterns. We chose five patterns as default. Generally, it is biologically meaningful with the number of patterns greater than 2. 
```{r}
nPatterns = 5
pdf(file = paste0("CommunicationPatterns_sender_heatmap.pdf"))
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
dev.off()
pdf(file = "patternAnalysis_sender_river.pdf", width = 7, height = 4)
netAnalysis_river(cellchat, pattern = "outgoing")
dev.off()
gg <- netAnalysis_dot(cellchat, pattern = "outgoing")
ggsave(filename="patternAnalysis_sender_dot.pdf", plot=gg, width = 5.5, height = 4, units = 'in', dpi = 300)
```

### Identify incoming communication pattern of target cells
```{r}
pdf(file = paste0("CommunicationPatterns_receiver_heatmap.pdf"))
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = nPatterns)
dev.off()
pdf(file = "patternAnalysis_receiver_river.pdf", width = 7, height = 4)
netAnalysis_river(cellchat, pattern = "incoming")
dev.off()
gg <- netAnalysis_dot(cellchat, pattern = "incoming")
ggsave(filename="patternAnalysis_receiver_dot.pdf", plot=gg, width = 5.5, height = 4, units = 'in', dpi = 300)
```


## Manifold and classification learning analysis
Further, CellChat is able to quantify the similarity between all significant signaling pathways and then group them based on their cellular communication network similarity. Grouping can be done either based on the functional or structural similarity. 

**Functional similarity**: High degree of functional similarity indicates major senders and receivers are similar, and it can be interpreted as the two signaling pathways or two ligand-receptor pairs exhibit similar and/or redundant roles. 

**Structural similarity**: A structural similarity was used to compare their signaling network structure, without considering the similarity of senders and receivers.

### Identify signaling groups based on their functional similarity
```{r}
cellchat <- computeNetSimilarity(cellchat, type = "functional", thresh = 0.25)
cellchat <- netEmbedding(cellchat, type = "functional")
cellchat <- netClustering(cellchat, type = "functional", k = 4)
gg <- netVisual_embedding(cellchat, type = "functional", pathway.remove.show = F)
cowplot::save_plot("2Dmanifold_FunctionalSimilarity_signalingPathways.pdf", gg, base_height = 3, base_width = 4)
pdf(file = "2Dmanifold_FunctionalSimilarity_signalingPathways_zoomIn.pdf", width = 2, height = 2.5*3)
netVisual_embeddingZoomIn(cellchat, type = "functional")
dev.off()
```


### Identify signaling groups based on structure similarity
```{r}
cellchat <- computeNetSimilarity(cellchat, type = "structural", thresh = 0.25)
cellchat <- netEmbedding(cellchat, type = "structural")
cellchat <- netClustering(cellchat, type = "structural")
gg <- netVisual_embedding(cellchat, type = "structural")
cowplot::save_plot("2Dmanifold_StructureSimilarity_signalingPathways.pdf", gg, base_height = 3, base_width = 4)
pdf(file = "2Dmanifold_StructureSimilarity_signalingPathways_zoomIn.pdf", width = 2, height = 2.5*3)
netVisual_embeddingZoomIn(cellchat, type = "structural")
dev.off()
```


## Save object
```{r}
saveRDS(cellchat, file = "cellchat_wound.rds")
```
